File: src/parsing/expressions/subquery_expression.ts
import ASTNode from "../ast_node";
import Operation from "../operations/operation";

/**
 * Counts rows flowing through an operation chain without producing results.
 * Used by SubqueryExpression to evaluate MATCH-only subqueries (no RETURN).
 */
class RowCounter extends Operation {
    public count: number = 0;
    public async run(): Promise<void> {
        this.count++;
    }
}

/**
 * Base class for the brace-delimited subquery expressions `EXISTS { }`,
 * `COUNT { }` and `COLLECT { }`.  The base executes the nested query and
 * hands the resulting rows and row-count to `reduce()`, which each concrete
 * subclass implements to produce its value.
 */
abstract class SubqueryExpression extends ASTNode {
    private _subqueryAST: ASTNode;
    private _value: any;

    constructor(subqueryAST: ASTNode) {
        super();
        this._subqueryAST = subqueryAST;
    }

    public introducesScope(): boolean {
        return true;
    }

    public async evaluate(): Promise<void> {
        // Imported lazily to break the module cycle with Runner (which
        // imports the parser, which imports this file's subclasses).
        const { default: Runner } = await import("../../compute/runner");
        const { default: Return } = await import("../operations/return");

        const first = this._subqueryAST.firstChild() as Operation;
        const last = this._subqueryAST.lastChild() as Operation;

        let rows: any[] = [];
        let count = 0;

        if (last instanceof Return) {
            const runner = new Runner(null, this._subqueryAST);
            await runner.run();
            rows = runner.results ?? [];
            count = rows.length;
        } else {
            // Subquery without RETURN (e.g., EXISTS { MATCH ... })
            const counter = new RowCounter();
            const savedNext = last.next;
            last.next = counter;

            await first.initialize();
            await first.run();
            await first.finish();

            last.next = savedNext;
            count = counter.count;
        }

        this._value = this.reduce(rows, count);
    }

    public value(): any {
        return this._value;
    }

    /** Reduce the executed subquery's rows and row-count to this expression's value. */
    protected abstract reduce(rows: any[], count: number): any;
}

export default SubqueryExpression;

File: flowquery-py/src/parsing/expressions/subquery_expression.py
"""Base class for EXISTS, COUNT, and COLLECT subquery expressions."""

from typing import Any, List

from ..ast_node import ASTNode
from ..operations.operation import Operation


class RowCounter(Operation):
    """Counts rows flowing through an operation chain without producing results.

    Used by SubqueryExpression to evaluate MATCH-only subqueries (no RETURN).
    """

    def __init__(self) -> None:
        super().__init__()
        self.count: int = 0

    async def run(self) -> None:
        self.count += 1


class SubqueryExpression(ASTNode):
    """Base class for the brace-delimited subquery expressions.

    The base executes the nested query and hands the resulting rows and
    row-count to `reduce()`, which each concrete subclass (ExistsSubquery,
    CountSubquery, CollectSubquery) implements to produce its value.
    """

    def __init__(self, subquery_ast: ASTNode) -> None:
        super().__init__()
        self._subquery_ast = subquery_ast
        self._value: Any = None

    def introduces_scope(self) -> bool:
        return True

    async def evaluate(self) -> None:
        from ...compute.runner import Runner
        from ..operations.return_op import Return

        first = self._subquery_ast.first_child()
        last = self._subquery_ast.last_child()

        if not isinstance(first, Operation) or not isinstance(last, Operation):
            raise ValueError("Subquery AST must contain Operations")

        rows: List[Any] = []
        count = 0

        if isinstance(last, Return):
            runner = Runner(ast=self._subquery_ast)
            await runner.run()
            rows = runner.results or []
            count = len(rows)
        else:
            # Subquery without RETURN (e.g., EXISTS { MATCH ... })
            counter = RowCounter()
            saved_next = last.next
            last.next = counter

            await first.initialize()
            await first.run()
            await first.finish()

            last.next = saved_next
            count = counter.count

        self._value = self.reduce(rows, count)

    def value(self) -> Any:
        return self._value

    def reduce(self, rows: List[Any], count: int) -> Any:
        raise NotImplementedError
